### PR TITLE
Implement Unison.Parsers, and rewrite tests to utilize

### PR DESCRIPTION
**Pull request by refried**
Main definitions are:
```haskell
unsafeParseType' :: [(V, Type V)] -> String -> Type V
unsafeParseTerm' :: [(V, Term V)] -> [(V, Type V)] -> String -> Term V
```
These still leave free variables for items not in the provided substitution lists. We may decide to change this later (you already told me to), but it was convenient for now, because Reasons.

These convenience functions use a default substitution whitelist defined in the same file
```haskell
unsafeParseTerm :: String -> Term V 
unsafeParseType :: String -> Type V
```

Tests are in `Unison.Test.TypeParser` and `Unison.Test.TermParser`

Moved Parser.hs from `editor/` to `shared/`
Updated to use `unsafeParse{Term,Type}`:
```haskell
Unison.Runtime.ExtraBuiltins
Unison.Node.BasicNode
Unison.Node.Builtin
Unison.Test.Term
Unison.Test.Typechecker
```

### CODE REVIEW

**Comment by refried** [comment]
I did a little bit on hash literals; thought I'd maybe finished it, but had blown something up in the process. I probably won't look again immediately.

**Comment by pchiusano** [comment]
Yay!! I probably won't get a chance to review in detail until Monday.

**Comment by refried** [comment]
@pchiusano P.S. happy to squash these

File: shared/src/Unison/TermParser.hs [hunk 1]
```diff
@@ -0,0 +1,239 @@
...
+
+term4 :: Parser (Term V)
+term4 = prefixApp term5
+
+term5 :: Parser (Term V)
+term5 = lam term <|> termLeaf
+
+termLeaf :: Parser (Term V)
+termLeaf = asum [prefixTerm, lit, parenthesized term, blank, vector term]
+
```
Comment (pchiusano) [inline shared/src/Unison/TermParser.hs]: These are really more notes to self that were useful when you were writing. I'd probably delete and just have as a note at the top:

> Precedence of language constructs is identical to Haskell, except that all operators (like `+`, `<*>`, or any sequence of non-alphanumeric characters) are left-associative and equal precedence, and operators must have surrounding whitespace (`a + b`, not `a+b`) to distinguish from identifiers that may contain operator characters (like `empty?` or `fold-left`).

This is economical and unambiguous.

File: shared/tests/Unison/Test/Term.hs [hunk 2]
```diff
@@ -109,40 +105,18 @@ rect x y w h =
...
-
 pingpong1 :: TTerm
 pingpong1 =
-  letRec'
-    [ ("ping", lam' ["x"] $ var' "pong" `app` (plus (var' "x") one))
-    , ("pong", lam' ["y"] $ var' "pong" `app` (minus (var' "y") one)) ]
-    (var' "ping" `app` one)
+  unsafeParseTerm $
+    unlines [ "let rec ping = x -> pong (x + 1)"
+            , "      ; pong = y -> ping (y - 1)"
```
Comment (pchiusano) [inline shared/tests/Unison/Test/Term.hs]: Ha, I kind of like this semicolon placement at the beginning of the next line.

File: shared/tests/Unison/Test/TermParser.hs [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+      Succeed a _ _ -> assertEqual "mismatch" expected a
+
+parseFail :: String -> TestTree
+parseFail s =
+  testCase ("`" ++ s ++ "` shouldn't parse") $ assertBool "should not have parsed" $
+    case parseTerm s of Fail {} -> True; Succeed {} -> False
+
+tests :: TestTree
+tests = testGroup "TermParser" $ (parse <$> shouldPass) ++ (parseFail <$> shouldFail)
+  where
```
Comment (pchiusano) [inline shared/tests/Unison/Test/TermParser.hs]: looks good

**Comment by pchiusano** [comment]
@refried looks good, merging. Can you fix up my one comment when you get a chance?

Also if you can get hash literals working that would be cool. You can do that as a separate PR.

**Pull request merged**